weblate: init at 5.6.2, add module

erictapen opened the pull request:
## Description of changes

This is (afaik) the third attempt to get Weblate into Nixpkgs (https://github.com/NixOS/nixpkgs/pull/48726, https://github.com/NixOS/nixpkgs/pull/169797) and is a combination of these efforts, plus my work on a separate flake https://github.com/ngi-nix/weblate which I did for Summer of Nix back then.

Marked as a draft for now, as I might need to file some changes in here as a separate PR. Also the commit structure is kinda messy and is grown over time, I'd do a cleaner rebase once it's ready for review. The project builds though and the NixOS tests run successully.

@jtojnar I included you as maintainer in a few packages you started in https://github.com/NixOS/nixpkgs/pull/48726, is that okay?

Extra thanks to @summersamara, who helped me through some python madness :heart: .

Thanks for Summer of Nix to fund this work!

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wegank commented on pkgs/development/python-modules/phply/default.nix:
```diff
@@ -0,0 +1,33 @@
+
+  meta = with lib; {
+    description = "Lexer and parser for PHP source implemented using PLY";
+    homepage = https://github.com/viraptor/phply;
```
```suggestion
    homepage = "https://github.com/viraptor/phply";
```

wegank commented on pkgs/development/python-modules/translate-toolkit/default.nix:
```diff
@@ -0,0 +1,73 @@
+
+  meta = with lib; {
+    description = "Useful localization tools for building localization & translation systems";
+    homepage = https://toolkit.translatehouse.org/;
```
```suggestion
    homepage = "https://toolkit.translatehouse.org/";
```

wegank commented on pkgs/development/python-modules/django-cors-headers/default.nix:
```diff
@@ -2,30 +2,34 @@
   src = fetchFromGitHub {
     owner = "adamchainz";
-    repo = "django-cors-headers";
+    repo = pname;
```
I believe `repo = "django-cors-headers";` is still preferred, athough the two have no difference under `buildPythonPackage rec`.

erictapen replied:
> You can also send changes to existing packages in separate PRs for faster reviews and merges.

Makes sense, I'm working on that

> I believe repo = "django-cors-headers"; is still preferred, athough the two have no difference under buildPythonPackage rec.

I see this pattern in more than four thousand places all over Nixpkgs, are you sure? Not important for me though.

wegank replied:
Not important to me either.

GetPsyched replied:
See https://github.com/nix-community/nixpkgs-lint/issues/21

wegank commented on pkgs/development/python-modules/aeidon/default.nix:
```diff
@@ -0,0 +1,48 @@
+  meta = with lib; {
+    description = "Reading, writing and manipulationg text-based subtitle files";
+    homepage = "https://github.com/otsaloma/gaupol";
+    license = licenses.gpl3;
```
`licenses.gpl3` is deprecated in favour of `licenses.gpl3Plus` and `licenses.gpl3Only`.

wegank commented on pkgs/development/python-modules/django-timezone-field/default.nix:
```diff
@@ -14,26 +14,25 @@
   ];
 
-  pythonImportsCheck = [ "timezone_field" ];
+  # pythonImportsCheck = [ "timezone_field" ];
```
Why is this commented?

wegank replied:
I think that check's controlled by `dontUsePythonImportsCheck` instead. Is the imports check failing on master?

erictapen replied:
No, the import check succeeds on master. I moved the change here for easier review: https://github.com/NixOS/nixpkgs/pull/325608

wegank commented on pkgs/development/python-modules/phply/default.nix:
```diff
@@ -0,0 +1,33 @@
+  meta = with lib; {
+    description = "Lexer and parser for PHP source implemented using PLY";
+    homepage = "https://github.com/viraptor/phply";
+    license = licenses.gpl2;
```
```suggestion
    license = licenses.bsd3;
```

https://github.com/viraptor/phply/blob/master/LICENSE looks more like BSD, I guess?

erictapen replied:
Oops, this was an oversight :grimacing: 
Looks like licenses.bsd3 to me, do you agree?

wegank replied:
I agree.

wegank commented on pkgs/development/python-modules/python-fluent/fluent-syntax.nix:
```diff
@@ -0,0 +1,39 @@
+buildPythonPackage {
+  pname = "fluent-syntax";
+  inherit version;
+  format = "setuptools";
```
Does `pyproject = true` work?

erictapen replied:
From my understanding it won't because their pyproject.toml doesn't define any information about the build system. But I never tested it.

GetPsyched replied:
It will not work since the project doesn't have `pyproject.toml` in the latest release tag.

See https://github.com/NixOS/nixpkgs/pull/255937#discussion_r1659674323

wegank replied:
In that case, setting `pyproject = true` is actually OK:

https://github.com/NixOS/nixpkgs/blob/d41a79d80befd2d5f57d4964f31a8e50171437df/doc/languages-frameworks/python.section.md?plain=1#L167-L169

GetPsyched replied:
Interesting, I'll update my PR in some time.

wegank commented on pkgs/top-level/all-packages.nix:
```diff
@@ -4007,6 +4007,8 @@ with pkgs;
 
   ios-webkit-debug-proxy = callPackage ../development/mobile/ios-webkit-debug-proxy { };
 
+  weblate = callPackage ../servers/web-apps/weblate { };
```
I guess `pkgs/servers/web-apps/weblate/default.nix` should be moved to `pkgs/by-name/we/weblate/package.nix` and this line should be removed.

erictapen replied:
Makes sense, pushed that.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-summer-of-nix-program-updates/46053/3

erictapen commented:
These PRs need to be merged for this PR to be ready for review:

- [ ] https://github.com/NixOS/nixpkgs/pull/255937
- [ ] https://github.com/NixOS/nixpkgs/pull/325598
- [ ] https://github.com/NixOS/nixpkgs/pull/325600
- [ ] https://github.com/NixOS/nixpkgs/pull/325604
- [ ] https://github.com/NixOS/nixpkgs/pull/325608

jtojnar commented:
> @jtojnar I included you as maintainer in a few packages you started in #48726, is that okay?

Sorry, I no longer have the capacity to maintain new packages.

wegank commented:
> As all blockers were merged, this is now ready for review. Should I squash the commits into a mergable set now or later? I figure there is some stuff that needs to be added to this.

I'd say now.

erictapen committed:
python3Packages.phply: init at 1.2.6

Co-authored-by: Jan Tojnar <jtojnar@gmail.com>
Co-authored-by: Weijia Wang <9713184+wegank@users.noreply.github.com>

erictapen committed:
python3Packages.aeidon: init at 1.15

erictapen committed:
python3Packages.translate-toolkit: init at 3.13.2

Co-authored-by: Jan Tojnar <jtojnar@gmail.com>
Co-authored-by: Weijia Wang <9713184+wegank@users.noreply.github.com>

erictapen committed:
python3Packages.ahocorasick-rs: init at 0.22.0

erictapen committed:
python3Packages.celery: add redis as optional dependency

erictapen committed:
python3Packages.django-crispy-bootstrap3: init at 2024.1

erictapen committed:
python3Packages.pyicumessageformat: init at 1.0.0

Co-authored-by: Taeer Bar-Yam <Radvendii@users.noreply.github.com>

erictapen committed:
python3Packages.siphashc: init at 2.4.1

Co-authored-by: Jan Tojnar <jtojnar@gmail.com>
Co-authored-by: Taeer Bar-Yam <Radvendii@users.noreply.github.com>

erictapen committed:
python3Packages.weblate-language-data: init at 2024.5

erictapen committed:
python3Packages.weblate-schemas: init at 2024.1

erictapen committed:
python3Packages.translation-finder: init at 2.16

erictapen committed:
weblate: init at 5.6.2

erictapen committed:
nixos/weblate: init module and test

Co-authored-by: Taeer Bar-Yam <Radvendii@users.noreply.github.com>

erictapen commented:
@wegank Done!

erictapen commented:
I just formatted every file with `nixfmt`, without adding further changes.

GetPsyched commented on pkgs/development/python-modules/pyicumessageformat/default.nix:
```diff
@@ -0,0 +1,31 @@
+  pythonImportsCheck = [ pname ];
+
+  meta = with lib; {
+    description = "An unopinionated Python3 parser for ICU MessageFormat";
```
```suggestion
    description = "Unopinionated Python3 parser for ICU MessageFormat";
```
https://github.com/NixOS/nixpkgs/blob/3836c3482484025821c9a55ca3db45c204ea5646/pkgs/README.md?plain=1#L378

erictapen replied:
Done

erictapen commented on pkgs/development/python-modules/translation-finder/default.nix:
```diff
@@ -0,0 +1,41 @@
+  version = "2.16";
+
+  src = fetchPypi {
+    inherit pname version;
```
Done, in all files :+1:

erictapen commented:
Adressed a merge conflict and fixed the new linting errors

Radvendii commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+          "BACKEND": "django_redis.cache.RedisCache",
+          "LOCATION": "unix://${config.services.redis.servers.weblate.unixSocket}",
+          "OPTIONS": {
+              "CLIENT_CLASS": "django_redis.client.DefaultClient",
```
Maybe this is no longer necessary, but in the flake you had `PARSER_CLASS`. I wonder if you ran into the same error relating to that I did. It resolved when I prefixed `HiredisParser` with a `_`.
```suggestion
              "CLIENT_CLASS": "django_redis.client.DefaultClient",
              "PARSER_CLASS": "redis.connection._HiredisParser",
```

Radvendii reviewed (commented):
Oh! We should have synced better :woman_facepalming: after [our exchange](https://github.com/ngi-nix/weblate/issues/4), I went ahead and tried to package it myself in much the same way you have.

I would have asked for your help polishing it up anyways, though; I'm not super familiar with python packaging.

I looked through the differences between our implementations and left comments about what I saw.

erictapen commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+  finalPackage = cfg.package.overridePythonAttrs (old: {
+    # We only support the PostgreSQL backend in this module
+    dependencies = old.dependencies ++ cfg.package.optional-dependencies.postgres;
+    # Use a settings module in dataDir, to avoid having to rebuild the package
+    # when user changes settings.
+    makeWrapperArgs = (old.makeWrapperArgs or [ ]) ++ [
+      "--set PYTHONPATH  \"${settingsDir}\""
+      "--set DJANGO_SETTINGS_MODULE \"settings\""
+    ];
```
I think the wording is not great here. The settings aren't mutable actually, they are just symlinked into `/var/lib`, so that we don't have a `/nix/store` reference to the settings in the weblate closure. Changing the module configuration will always deploy and restart the right settings. There wouldn't be any state.

I'll refine the comment.

Radvendii replied:
Ohh I see.
1. Why are we doing the same with `staticRoot` / `compressRoot`
2. Given that we run this through uwsgi, are the wrapper env variables even getting used anywhere?

erictapen replied:
1. I realised that `staticRoot`/`compressRoot` where remnants of before I figured out to use a `settingsDir`. I removed these!
2. Yes, when one creates the first admin user for example, using the `weblate` executable in path. In the flake, I used weblate-env for this, which is now unnecessary. Also the migration service calls weblate directly.

Radvendii replied:
How do you feel about adding another layer of wrapper instead? Something like
```
wrappedWeblate = pkgs.runCommand "wrap-weblate" { nativeBuildInputs = [ pkgs.makeWrapper ]; } ''
  mkdir -p $out/bin
  makeWrapper ${cfg.package}/bin/weblate $out/bin/weblate \
    --set PYTHONPATH "${settings_py}"
    --set DJANGO_SETTINGS_MODULE "settings"
'';
```
It trades complexity in one place for complexity somewhere else, so I'll leave the judgment call to you. Personally `tmpfiles.d` and symlinking store paths as runtime paths feels like icky state and therefore a last resort. It's not actually so bad in this case, though.

erictapen replied:
I thought about it a bit, but I think I'd like to leave it like it is. Personally I sometimes take comfort in knowing a semi-static location for the service configuration (compare to having to peel the path out of the systemd service definition every time the config changes) and I don't like the idea of wrapping all weblate binaries so much, even though I understand that it'd totally work to wrap the binary twice.

erictapen commented on pkgs/development/python-modules/translation-finder/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  pythonImportsCheck = [ "translation_finder" ];
+
+  doCheck = false;
```
Oops, that was indeed an oversight!

erictapen commented:
Thanks for the thorough review. I'll reply to it in the coming day and push the fixups for now and squash once everything is addressed.

erictapen commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+  environment = {
+    PYTHONPATH = "${settingsDir}:${pythonEnv}/${python.sitePackages}/";
+    DJANGO_SETTINGS_MODULE = "settings";
+    inherit (finalPackage) GI_TYPELIB_PATH;
```
Good idea, done.

erictapen commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+      "L+ ${staticRoot} - - - - ${finalPackage.static}/static"
+      "L+ ${compressRoot} - - - - ${finalPackage.static}/compressor-cache"
+      "L+ ${settingsDir} - - - - ${settings_py}"
+    ];
```
I remove staticRoot and compressRoot, these where indeed unnecessary remnants.

erictapen commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+      serviceConfig = {
+        Type = "notify";
+        NotifyAccess = "all";
+        ExecStart =
+          let
+            uwsgi = pkgs.uwsgi.override {
+              plugins = [ "python3" ];
+              python3 = python;
+            };
+            jsonConfig = pkgs.writeText "uwsgi.json" (builtins.toJSON uwsgiConfig);
+          in
+          "${uwsgi}/bin/uwsgi --json ${jsonConfig}";
+        Restart = "on-failure";
+        KillSignal = "SIGTERM";
+        WorkingDirectory = finalPackage;
```
[Upstreams docs](https://docs.weblate.org/en/latest/admin/install.html#uwsgi) seem to tie uwsgi to Ngnix and gunicorn to Apache, even though it obviously seems to run differently. For now I think I'd rather follow upstreams setup recommendations with Nginx/uwsgi as closely, to be confident in the configuration.

I had my struggles with uwsgi, but didn't notice any problems recently.

erictapen replied:
I'm reconsidering using unicorn. Could you share your full code with me? I think the snippet doesn't contain how you expose the `uwsgi` module to gunicorn.

Radvendii replied:
I think that comes from `weblate/wsgi.py` in the weblate repository. You just have to set the `WorkingDirectory` appropriately. Not sure if that's still `dataDir` or if it would be `${finalPackage}/${python.sitePackages}/weblate/`

erictapen replied:
That worked, thanks!

erictapen commented on nixos/tests/web-apps/weblate.nix:
```diff
@@ -0,0 +1,121 @@
+          enable = true;
+          localDomain = "${serverDomain}";
+          djangoSecretKeyFile = pkgs.writeText "weblate-django-secret" "thisissnakeoilsecretwithmorethan50characterscorrecthorsebatterystaple";
+          # smtp = {
+          #   enable = true;
+          #   host = "localhost";
+          #   user = "weblate@${serverDomain}";
+          #   passwordFile = pkgs.writeText "weblate-smtp-pass" "thisissnakeoilpassword";
+          # };
```
I figured they might work as a pathfinder for future developers (e.g. me) who want to enable email sending in the test to make all the deploy checks green. But probably not working enough to be kept as such, I'll remove the comments.

erictapen commented on pkgs/by-name/we/weblate/package.nix:
```diff
@@ -0,0 +1,193 @@
+      celery = prev.celery.overridePythonAttrs (old: {
+        dependencies = old.dependencies ++ prev.celery.optional-dependencies.redis;
+      });
+      python-redis-lock = prev.python-redis-lock.overridePythonAttrs (old: {
+        dependencies = old.dependencies ++ prev.python-redis-lock.optional-dependencies.django;
+      });
```
python-redis-lock actually uses `optional-dependencies` since https://github.com/NixOS/nixpkgs/pull/325600.

erictapen commented on pkgs/by-name/we/weblate/package.nix:
```diff
@@ -0,0 +1,193 @@
+  pname = "weblate";
+  version = "5.6.2";
+
+  pyproject = true;
```
I'm pretty sure it's the former. I saw it everywhere and I read it's supposed to become the default, with other formats becoming fallbacks.

erictapen commented on pkgs/by-name/we/weblate/package.nix:
```diff
@@ -0,0 +1,193 @@
+    gdk-pixbuf
+    glib
+  ];
+  makeWrapperArgs = [ "--set GI_TYPELIB_PATH \"$GI_TYPELIB_PATH\"" ];
```
This is actually redundant, so I removed the use of `wrapGappsNoGuiHook`. We need to specify `GI_TYPELIB_PATH` explicitly, so that we can expose it through `passthru`. That's the only reason. I added a comment to explain this.

erictapen commented on pkgs/development/python-modules/phply/default.nix:
```diff
@@ -0,0 +1,31 @@
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  pythonImportsCheck = [ "phply" ];
```
So in an earlier review I went through all code and removed stuff like `fetchFromGitHub { repo = pname; .. }`. The rationale behind it actually makes sense to, even though it's a super widely used pattern: Changing the `pname` is unlikely to be correlated with a change in the repo location. I'd assume the same about the python module name.

Radvendii replied:
Ah. You missed some other ones below. I guessed wrong which was the intended format.

erictapen replied:
Thanks, done.

erictapen commented on pkgs/by-name/we/weblate/package.nix:
```diff
@@ -0,0 +1,193 @@
+  patches = [
+    # FIXME This shouldn't be necessary and probably has to do with some dependency mismatch.
+    ./cache.lock.patch
+    # relax dependency constraint for borgbackup
+    (fetchpatch {
+      url = "https://github.com/WeblateOrg/weblate/commit/355c81c977c59948535a98a35a5c05d7e6909703.patch";
+      sha256 = "sha256-9cgE+KeSsXweb2JUvstMxEehfNVktZm1geAo496EPN0=";
+    })
+    # relax dependency constraint for django-crispy-forms
+    (fetchpatch {
+      url = "https://github.com/WeblateOrg/weblate/commit/7b341c523ed9b3b41ecfbc5c92dd6156992e4f32.patch";
+      sha256 = "sha256-AcWQD+Z1wGX9MT3IFiomN4sncsQ2VOtflo6BLRZbxBc=";
+    })
+  ];
+
+  # relax dependency constraint for mistletoe
+  # https://github.com/WeblateOrg/weblate/commit/50df46a25dda2b7b39de86d4c65ecd7a685f62e6
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace '"mistletoe>=1.3.0,<1.4"' '"mistletoe>=1.3.0,<1.5"'
+  '';
```
This certainly would be less work, but obviously produces nastier build (or runtime) errors. I don't really have a taste in this, just made the experience in the past that people here preferred more constraints than less (though tbf that was in the `haskellPackages` set). If you'd prefer `pythonRelaxDeps` I'd use that, just give me the call.

Radvendii replied:
I don't have a preference myself. Happy to leave it the way you have it now.

erictapen commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+      path = weblatePath;
+      serviceConfig = {
+        Type = "oneshot";
+        # WorkingDirectory = finalPackage;
```
Apparently not. I'll remove the commented code. I also just checked, `StateDir` is required actually because `migrate` creates directories.

erictapen commented on nixos/tests/web-apps/weblate.nix:
```diff
@@ -0,0 +1,121 @@
+        #   # sslKey = certs."${serverDomain}".key;
+        #   # sslCert = certs."${serverDomain}".cert;
+        # };
+
```
done

Radvendii commented on pkgs/development/python-modules/pyicumessageformat/default.nix:
```diff
@@ -0,0 +1,32 @@
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  pythonImportsCheck = [ pname ];
```
```suggestion
  pythonImportsCheck = [ "pyicumessageformat" ];
```

Radvendii commented on pkgs/development/python-modules/siphashc/default.nix:
```diff
@@ -0,0 +1,31 @@
+
+  nativeCheckInputs = [ pytestCheckHook ];
+
+  pythonImportsCheck = [ pname ];
```
```suggestion
  pythonImportsCheck = [ "siphashc" ];
```

Radvendii commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,421 @@
+
+    ''
+    + cfg.extraConfig;
+  settings_py = pkgs.runCommand "weblate_settings.py" { } ''
+    mkdir -p $out
+    cat ${finalPackage}/${python.sitePackages}/weblate/settings_example.py > $out/settings.py
+    cat >> $out/settings.py <<EOF
+    ${weblateConfig}EOF
+  '';
```
Eh. In some ways this is nicer, but it uses less well-known nixpkgs features. I'm not attached to this suggestion but thought it was worth noting that it's available in case you like it.
```suggestion
  settings_py = pkgs.runCommand "weblate_settings.py" {
    inherit weblateConfig;
    passAsFile = [ "weblateConfig" ];
  } ''
    mkdir -p $out
    cat \
      ${finalPackage}/${python.sitePackages}/weblate/settings_example.py \
      $weblateConfigPath \
      > $out/settings.py
  '';
```

erictapen replied:
Love it, I always prefer nixpkgs features over bash features.

erictapen commented:
@Radvendii I think I addressed now all of your review points (except for https://github.com/NixOS/nixpkgs/pull/325541#discussion_r1697655995 maybe?). Would you mind having another look?

If you give the go, I'd sqash again and I feel like this might be close to mergeable after. @fricklerhandwerk

Radvendii commented on nixos/modules/services/web-apps/weblate.nix:
```diff
@@ -0,0 +1,422 @@
+          $weblateConfigPath \
+          > $out/settings.py
+      '';
+  uwsgiConfig.uwsgi = {
+    type = "normal";
+    plugins = [ "python3" ];
+    master = true;
+    socket = "/run/weblate.socket";
+    die-on-idle = true;
+    die-on-term = true;
+    idle = 600;
+    manage-script-name = true;
+    cheap = true;
+    chmod-socket = "770";
+    chown-socket = "weblate:weblate";
+    uid = "weblate";
+    gid = "weblate";
+    wsgi-file = "${finalPackage}/${python.sitePackages}/weblate/wsgi.py";
+
+    # Some more recommendations by upstream:
+    # https://docs.weblate.org/en/latest/admin/install.html#sample-configuration-for-nginx-and-uwsgi
+    buffer-size = 8192;
+    reload-on-rss = 250;
+    workers = 8;
+    enable-threads = true;
+    close-on-exec = true;
+    umask = "0022";
+    ignore-sigpipe = true;
+    ignore-write-errors = true;
+    disable-write-exception = true;
+  };
```
Is this basically **the** configuration to be used with weblate, or would people want to change this when they configure the module? Right now it's quite a pain to override.

erictapen replied:
Actually only the lower part after the comment are actual recommendations by upstream, the upper part is mostly stuff that I set and then forgot about why I did it :see_no_evil: 

I agree with your concerns about overridability. I also looked a bit into uwsgi and realised the project is in maintenance mode for some time now. Maybe it just makes sense to switch to gunicorn, as you proposed [here](https://github.com/NixOS/nixpkgs/pull/325541#discussion_r1695185961).

erictapen replied:
Using gunicorn instead worked great!

Radvendii commented on pkgs/by-name/we/weblate/package.nix:
```diff
@@ -0,0 +1,187 @@
+    # FIXME This shouldn't be necessary and probably has to do with some dependency mismatch.
+    ./cache.lock.patch
+    # relax dependency constraint for borgbackup
+    (fetchpatch {
+      url = "https://github.com/WeblateOrg/weblate/commit/355c81c977c59948535a98a35a5c05d7e6909703.patch";
+      sha256 = "sha256-9cgE+KeSsXweb2JUvstMxEehfNVktZm1geAo496EPN0=";
+    })
+    # relax dependency constraint for django-crispy-forms
+    (fetchpatch {
+      url = "https://github.com/WeblateOrg/weblate/commit/7b341c523ed9b3b41ecfbc5c92dd6156992e4f32.patch";
+      sha256 = "sha256-AcWQD+Z1wGX9MT3IFiomN4sncsQ2VOtflo6BLRZbxBc=";
+    })
+  ];
+
+  # relax dependency constraint for mistletoe
+  # https://github.com/WeblateOrg/weblate/commit/50df46a25dda2b7b39de86d4c65ecd7a685f62e6
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace '"mistletoe>=1.3.0,<1.4"' '"mistletoe>=1.3.0,<1.5"'
+  '';
```
Is there a reason mistletoe is being relaxed in a different way than borgbackup and django-crispy-forms? Does that patch not apply?

Maybe we can switch them all to the `substituteInPlace` method. It makes it clear what's being updated, and makes it easier to update in the future. Do we need to be careful to only update things that have been updated upstream? Does that even guarantee us anything, given that something might have changed upstream in the meantime to add compatibility with that version?

erictapen replied:
Yeah, the specific patch didn't apply. I liked the idea of having an easy way to verify that the change was also done by upstream, but ofc it doesn't guarantee anything as you said. I'll switch to `substituteInPlace` and add comments to the respective commits.

Radvendii reviewed (approved):
Two minor comments / concerns / questions, but overall this looks good to me.

erictapen commented:
@Radvendii @GetPsyched @wegank I just reworked the systemd ordering in the latest commit, as I felt I could remove some redundant code here. Do you mind having another look at that change, to make sure I didn't do something weird there? Keep in mind Weblate is socket activated as is, though celery and the migrations run from boot.

erictapen commented:
@GrahamcOfBorg build weblate

fricklerhandwerk commented:
Thanks a lot everyone!

dotlambda commented on pkgs/by-name/we/weblate/package.nix:
```diff
@@ -0,0 +1,180 @@
+  dependencies = with python.pkgs; [
+    aeidon
+    ahocorasick-rs
+    borgbackup
```
This doesn't make any sense. Weblate doesn't use `import borg`. You should either use `makeWrapper` to prefix `$PATH` or hardcode the path to the executable in https://github.com/WeblateOrg/weblate/blob/weblate-5.7/weblate/utils/backup.py#L85

dotlambda replied:
Another alternative would be removing the dependency outright and adding borgbackup to the module's `$PATH`.

erictapen replied:
> This doesn't make any sense. Weblate doesn't use import borg. 

Oh it does: https://github.com/WeblateOrg/weblate/blob/weblate-5.7/weblate/utils/validators.py#L231
borgbackups python library is also explicitly listed as a dependency in their pyproject.toml, that's why I added it.

Though it looks like weblate indeed needs borgs executable in PATH, so I'll add that to the module.

dotlambda replied:
> borgbackups python library

It's not a Python library. You didn't even use `toPythonModule`.

dotlambda replied:
I suggest you patch `validate_backup_path()` to call the executable instead.

erictapen replied:
Can you please address my points?

- borg *is* imported in Weblate code
- it is explicitly required as a poetry dependency
- Not mentioned before, but Weblate literally fails to build when I remove borgbackup from dependencies

So is using `(toPythonModule borgbackup)` in dependencies a solution you can live with?

Also I'm not going to patch the source file. Weblate requires borgbackup to be available, so I'll provide it there: Both in the wrapper args and in the systemd service path. Otherwise this is going to break easily in the future when they call the executable in a second place.

dotlambda replied:
> So is using `(toPythonModule borgbackup)` in dependencies a solution you can live with?

Yes.

dotlambda replied:
> * borg _is_ imported in Weblate code
> 
> * it is explicitly required as a poetry dependency

I don't think that's a very smart decision on upstream's part because borgbackup's internal API might change at any point.

dotlambda commented on pkgs/development/python-modules/aeidon/default.nix:
```diff
@@ -0,0 +1,61 @@
+buildPythonPackage rec {
+  pname = "aeidon";
+  version = "1.15";
+
```
```suggestion
  pyproject = true;

```

dotlambda commented on pkgs/development/python-modules/aeidon/default.nix:
```diff
@@ -0,0 +1,61 @@
+  src = fetchPypi {
+    pname = "aeidon";
+    inherit version;
+    sha256 = "sha256-qGpGraRZFVaW1Jys24qvfPo5WDg7Q/fhvm44JH8ulVw=";
```
```suggestion
    hash = "sha256-qGpGraRZFVaW1Jys24qvfPo5WDg7Q/fhvm44JH8ulVw=";
```

dotlambda commented on pkgs/development/python-modules/aeidon/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  installPhase = ''
+    runHook preInstall
+    python setup.py --without-gaupol install --prefix=$out
```
Use `setupPyBuildFlags`.

erictapen replied:
I had trouble doing that, as just adding `setupPyBuildFlags = [ "--without-gaupol" ]` results in a working aeidon build, but its reverse dependency `translate-toolkit` fails then due to missing directories. Any ideas?

dotlambda replied:
fixed in #338224

dotlambda commented on pkgs/development/python-modules/ahocorasick-rs/default.nix:
```diff
@@ -0,0 +1,53 @@
+  src = fetchPypi {
+    inherit version;
+    pname = "ahocorasick_rs";
+    sha256 = "sha256-lzRwODlJlymMSih3CqNIeR+HrUbgVhroM1JuHFfW848=";
```
```suggestion
    hash = "sha256-lzRwODlJlymMSih3CqNIeR+HrUbgVhroM1JuHFfW848=";
```

dotlambda commented on pkgs/development/python-modules/ahocorasick-rs/default.nix:
```diff
@@ -0,0 +1,53 @@
+  nativeBuildInputs = with rustPlatform; [
+    maturinBuildHook
+    cargoSetupHook
+    typing-extensions
```
If anywhere this should be in `dependencies`:
https://github.com/G-Research/ahocorasick_rs/blob/034e3f67e12198c08137bb9fb3153cb01cf5da31/pyproject.toml#L9-L10

dotlambda commented on pkgs/development/python-modules/django-crispy-bootstrap3/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+  dependencies = [
+    django
+    setuptools
```
That's the `build-system`!

dotlambda commented on pkgs/development/python-modules/django-crispy-bootstrap3/default.nix:
```diff
@@ -0,0 +1,43 @@
+  nativeCheckInputs = [
+    pytest-django
+    pytestCheckHook
+    django-crispy-forms
```
should be in `dependencies`

dotlambda commented on pkgs/development/python-modules/phply/default.nix:
```diff
@@ -0,0 +1,31 @@
+buildPythonPackage rec {
+  pname = "phply";
+  version = "1.2.6";
+
```
```suggestion
  pyproject = true;

```

dotlambda commented on pkgs/development/python-modules/pyicumessageformat/default.nix:
```diff
@@ -0,0 +1,32 @@
+  build-system = [ setuptools ];
+
+  src = fetchPypi {
+    pname = "pyicumessageformat";
+    inherit version;
```
```suggestion
    inherit pname version;
```

erictapen replied:
The motivation for this was https://github.com/nix-community/nixpkgs-lint/issues/21, it got addressed already in an earlier review.

dotlambda replied:
It's different for `fetchPypi`.

GetPsyched replied:
How is it different?

dotlambda replied:
`repo = pname` is accidental but the `pname` of a Python package is supposed to be exactly what it's called on PyPI.

dotlambda replied:
https://github.com/NixOS/nixpkgs/issues/277994#issuecomment-2318421051

erictapen replied:
Makes sense to me, I'll change that.

dotlambda commented on pkgs/development/python-modules/translate-toolkit/default.nix:
```diff
@@ -0,0 +1,72 @@
+  src = fetchPypi {
+    pname = "translate_toolkit";
+    inherit version;
+    sha256 = "sha256-95zIAelFSNK5+f1GY8DUgHPDQBS5K+9ULjXaSaa0wWM=";
```
```suggestion
    hash = "sha256-95zIAelFSNK5+f1GY8DUgHPDQBS5K+9ULjXaSaa0wWM=";
```

dotlambda commented on pkgs/development/python-modules/translation-finder/default.nix:
```diff
@@ -0,0 +1,40 @@
+  version = "2.16";
+
+  src = fetchPypi {
+    pname = "translation-finder";
+    inherit version;
```
```suggestion
    inherit pname version;
```

erictapen replied:
See https://github.com/NixOS/nixpkgs/pull/325541#discussion_r1736630328

dotlambda commented on pkgs/development/python-modules/translation-finder/default.nix:
```diff
@@ -0,0 +1,40 @@
+buildPythonPackage rec {
+  pname = "translation-finder";
+  version = "2.16";
+
```
```suggestion
  pyproject = true;

```

dotlambda commented on pkgs/development/python-modules/weblate-schemas/default.nix:
```diff
@@ -0,0 +1,40 @@
+buildPythonPackage rec {
+  pname = "weblate-schemas";
+  version = "2024.1";
+
```
```suggestion
  pyproject = true;

```

dotlambda commented:
Did anyone even look at the Python packages during review??

GetPsyched commented:
I did, though at the time I wasn't aware of the newer Python package conventions. (missing the `sha256` to `hash` was an oversight though)

erictapen commented:
@dotlambda I addressed your reviews in https://github.com/NixOS/nixpkgs/pull/338217, though I don't really get the strength of your tone. Looks like it was mostly nits, that can just be solved by better linting eventually.
